### PR TITLE
go/store/nbs: use chunk journal index to speedup bootstrapping

### PR DESCRIPTION
**Pull request by andy-wm-arthur**

### CODE REVIEW

File: go/store/nbs/journal_writer.go [hunk 1]
```diff
@@ -142,14 +149,72 @@ var _ io.Closer = &journalWriter{}
...
+				}
+				// populate range hashmap
+				for _, l := range deserializeLookups(r.payload) {
+					wr.ranges.put(l.a, l.r)
+				}
+				// record a high-water-mark for the indexed portion of the journal
+				wr.indexed = int64(r.end)
+				// todo: uncompressed size
+			default:
+				return fmt.Errorf("unknown index record kind (%d)", r.kind)
```
Comment (andy-wm-arthur) [inline go/store/nbs/journal_writer.go:196]: reading an unknown record kind will throw an error, causing the index to be deleted in `corruptIndexRecovery`... probably not the behavior we want

File: go/store/nbs/journal_writer.go [hunk 2]
```diff
@@ -142,14 +149,72 @@ var _ io.Closer = &journalWriter{}
...
+
+	if ok {
+		var info os.FileInfo
+		if info, err = wr.index.Stat(); err != nil {
+			return hash.Hash{}, err
+		}
+		err = processIndexRecords(ctx, wr.index, info.Size(), func(o int64, r indexRec) (err error) {
+			switch r.kind {
+			case tableIndexRecKind:
+				// |r.stop| is expected to point to a root hash record in |wr.journal|
```
Comment (reltuk) [inline go/store/nbs/journal_writer.go:180]:
```suggestion
				// |r.end| is expected to point to a root hash record in |wr.journal|
```

**Review by reltuk — APPROVED** [review]
Generally looks totally reasonable. A few comments.

File: go/store/nbs/journal_index_record.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+		}
+
+		l := readUint32(buf)
+		if buf, err = rdr.Peek(int(l)); err != nil {
+			break
+		}
+
+		// we do not zero-fill the journal index and expect
+		// only complete records that will checksum
+		if !validateIndexRecord(buf) {
```
[the remark this answers is not in the document]
Comment (andy-wm-arthur) [reply go/store/nbs/journal_index_record.go:203]: for now we're overly restrictive on the well-formedness of journal indexes. as you suggested, we could be more permissive or incrementally clean up the index, but for now we'll punt

**Pull request merged**